yarpl/include/yarpl/flowable/Subscriber.h

```diff
@@ -15,40 +15,118 @@ namespace flowable {
   // Note: If any of the following methods is overridden in a subclass, the new
   // methods SHOULD ensure that these are invoked as well.
-  virtual void onSubscribe(Reference<Subscription> subscription) {
+  void onSubscribe(Reference<Subscription> subscription) override {
     DCHECK(subscription);
-
-    if (subscription_) {
-      subscription->cancel();
-      return;
-    }
-
+    CHECK(!subscription_);
     subscription_ = std::move(subscription);
   }
 
   // No further calls to the subscription after this method is invoked.
-  virtual void onComplete() {
+  void onComplete() override {
     DCHECK(subscription_) << "Calling onComplete() without a subscription";
     subscription_.reset();
   }
 
   // No further calls to the subscription after this method is invoked.
-  virtual void onError(folly::exception_wrapper) {
+  void onError(folly::exception_wrapper) override {
     DCHECK(subscription_) << "Calling onError() without a subscription";
     subscription_.reset();
   }
 
-  virtual void onNext(T) = 0;
-
  protected:
-  Subscription* subscription() {
-    return subscription_.operator->();
+  Reference<Subscription> subscription() {
+    return subscription_;
   }
 
  private:
   Reference<Subscription> subscription_;
 };
+
+// T : Type of Flowable that this Subscriber operates on
+//
+// keep_reference_to_this : BaseSubscriber will keep a live reference to
+// itself on the stack while in a signaling or requesting method, in case
+// the derived class causes all other references to itself to be dropped.
+//
+// Classes that ensure that at least one reference will stay live can
+// use `keep_reference_to_this = false` as an optimization to
+// prevent an atomic inc/dec pair
+template <typename T>
+class BaseSubscriber : public Subscriber<T> {
+ public:
+  // Note: If any of the following methods is overridden in a subclass, the new
+  // methods SHOULD ensure that these are invoked as well.
+  void onSubscribe(Reference<Subscription> subscription) final override {
+    DCHECK(subscription);
+    CHECK(!subscription_);
+
+    subscription_ = std::move(subscription);
+    onSubscribeImpl();
+  }
+
+  // No further calls to the subscription after this method is invoked.
+  void onComplete() final override {
+    DCHECK(subscription_) << "Calling onComplete() without a subscription";
+
+    if(auto sub = subscription_.exchange(nullptr)) {
+      onCompleteImpl();
+      onTerminateImpl();
+    }
+  }
+
+  // No further calls to the subscription after this method is invoked.
+  void onError(folly::exception_wrapper e) final override {
+    DCHECK(subscription_) << "Calling onError() without a subscription";
+
+    if(auto sub = subscription_.exchange(nullptr)) {
+      onErrorImpl(std::move(e));
+      onTerminateImpl();
+    }
+  }
+
+  void onNext(T t) final override {
+    if(auto sub = subscription_.load()) {
+      onNextImpl(std::move(t));
+    }
+  }
+
+  void cancel() {
+    if(auto sub = subscription_.exchange(nullptr)) {
```

Comment:
I thought we would not have any reference to subscription_, but we would just call the cancel & request using a pure virtual method. @alexmalyshev, @lehecka